# my-perl
    You can find many short but practical PERL scripts about NGS data analysis in current directory !
# scripts list
*   1.calc_gene_exon_depth.pl
*   2.get_gene_from_chr_pos.pl
*   3.Degenerate_base2original_base.pl
*   4.snp_mutation_context.pl
*   5.stat_snp_subtype.pl
*   6.stat_WES_depth_covarage.pl

## 1.draw exon covarage map for each refseq transcript according a gene sysmbol  

